### PR TITLE
Update logging to use the blobstore service

### PR DESCRIPTION
**Pull request by damccorm**

### CODE REVIEW

File: src/Microsoft.VisualStudio.Services.Agent/JobServer.cs [hunk 1]
```diff
@@ -118,5 +125,49 @@ public Task<Timeline> GetTimelineAsync(Guid scopeIdentifier, string hubName, Gui
...
+        public async Task<BlobIdentifierWithBlocks> UploadLogToBlobstorageService(Stream blob, string hubName, Guid planId, int logId)
+        {
+            CheckConnection();
+
+            BlobIdentifierWithBlocks blobIdWithBlocks = VsoHash.CalculateBlobIdentifierWithBlocks(blob);
+            BlobIdentifier blobId = blobIdWithBlocks.BlobId;
+
+            using(var blobClient = CreateArtifactsClient(_connection, default(CancellationToken)))
+            {   
+                return await blobClient.UploadBlocksForBlobAsync(blobId, blob, default(CancellationToken));
```
Comment (damccorm) [inline src/Microsoft.VisualStudio.Services.Agent/JobServer.cs:157]: This call hangs indefinitely right now :(

**Comment by damccorm** [comment]
Note - CI is failing due to a package error unrelated to these changes (it also repros on master)

File: src/Microsoft.VisualStudio.Services.Agent/JobServerQueue.cs [hunk 2]
```diff
@@ -602,10 +609,26 @@ private async Task UploadFile(UploadFileInfo file)
...
 
-                    // Upload the contents
-                    using (FileStream fs = File.Open(file.Path, FileMode.Open, FileAccess.Read, FileShare.ReadWrite))
+                    if (_writeToBlobstorageService)
                     {
-                        var logUploaded = await _jobServer.AppendLogContentAsync(_scopeIdentifier, _hubName, _planId, taskLog.Id, fs, default(CancellationToken));
+                        BlobIdentifierWithBlocks blobBlockId = null;
+                        using (FileStream fs = File.Open(file.Path, FileMode.Open, FileAccess.Read, FileShare.ReadWrite))
+                        {
+                            blobBlockId = await _jobServer.UploadLogToBlobstorageService(fs, _hubName, _planId, taskLog.Id);
```
Comment (alex-peck) [inline src/Microsoft.VisualStudio.Services.Agent/JobServerQueue.cs:617]: Did we decide not to fallback to FCS on error?

Comment (damccorm) [reply src/Microsoft.VisualStudio.Services.Agent/JobServerQueue.cs]: Nope, I just forgot to do that - should be fixed

File: src/Microsoft.VisualStudio.Services.Agent/JobServerQueue.cs [hunk 3]
```diff
@@ -602,10 +609,32 @@ private async Task UploadFile(UploadFileInfo file)
...
+                                blobBlockId = await _jobServer.UploadLogToBlobstorageService(fs, _hubName, _planId, taskLog.Id);
+
+                                int lineCount = File.ReadLines(file.Path).Count();
+
+                                // Notify TFS
+                                await _jobServer.AssociateLogAsync(_scopeIdentifier, _hubName, _planId, taskLog.Id, blobBlockId, lineCount, default(CancellationToken));
+                            }
+                            catch {
+                                // Fall back to FCS
+                                await _jobServer.AppendLogContentAsync(_scopeIdentifier, _hubName, _planId, taskLog.Id, fs, default(CancellationToken));
```
Comment (alex-peck) [inline src/Microsoft.VisualStudio.Services.Agent/JobServerQueue.cs:627]: Do we need to reset the stream back to the beginning here? I'm not 100% on how streams work but if `UploadLogToBlobstoreService` reads some data and then fails we want to include all of it in this call.

File: src/Microsoft.VisualStudio.Services.Agent/JobServerQueue.cs [hunk 4]
```diff
@@ -602,10 +609,32 @@ private async Task UploadFile(UploadFileInfo file)
                     // Create the log
                     var taskLog = await _jobServer.CreateLogAsync(_scopeIdentifier, _hubName, _planId, new TaskLog(String.Format(@"logs\{0:D}", file.TimelineRecordId)), default(CancellationToken));
 
-                    // Upload the contents
-                    using (FileStream fs = File.Open(file.Path, FileMode.Open, FileAccess.Read, FileShare.ReadWrite))
+                    if (_writeToBlobstorageService)
                     {
-                        var logUploaded = await _jobServer.AppendLogContentAsync(_scopeIdentifier, _hubName, _planId, taskLog.Id, fs, default(CancellationToken));
+                        BlobIdentifierWithBlocks blobBlockId = null;
+                        using (FileStream fs = File.Open(file.Path, FileMode.Open, FileAccess.Read, FileShare.ReadWrite))
```
Comment (alex-peck) [inline src/Microsoft.VisualStudio.Services.Agent/JobServerQueue.cs:615]: nit: put this `using` before ` if (_writeToBlobstorageService)` to not duplicate it in both if and else

File: src/Microsoft.VisualStudio.Services.Agent/JobServerQueue.cs (hunk 3 above)
Comment (damccorm) [reply src/Microsoft.VisualStudio.Services.Agent/JobServerQueue.cs]: I think you're right

File: src/Microsoft.VisualStudio.Services.Agent/JobServerQueue.cs [hunk 5]
```diff
@@ -602,10 +609,34 @@ private async Task UploadFile(UploadFileInfo file)
...
+                                catch {
+                                    // Fall back to FCS
+                                    fs.Position = 0;
+                                    await _jobServer.AppendLogContentAsync(_scopeIdentifier, _hubName, _planId, taskLog.Id, fs, default(CancellationToken));
+                                }
+                        }
+                        else
+                        {
+                            // Upload the contents
+                            using (FileStream fs = File.Open(file.Path, FileMode.Open, FileAccess.Read, FileShare.ReadWrite))
```
Comment (alex-peck) [inline src/Microsoft.VisualStudio.Services.Agent/JobServerQueue.cs:635]: This isn't needed anymore

File: src/Microsoft.VisualStudio.Services.Agent/JobServer.cs [hunk 6]
```diff
@@ -118,5 +124,39 @@ public Task<Timeline> GetTimelineAsync(Guid scopeIdentifier, string hubName, Gui
...
+            {
+                return await blobClient.UploadBlocksForBlobAsync(blobId, blob, default(CancellationToken));
+            }
+        }
+
+        private IBlobStoreHttpClient CreateArtifactsClient(VssConnection connection, CancellationToken cancellationToken){
+            var tracer = new CallbackAppTraceSource(str => Trace.Info(str), System.Diagnostics.SourceLevels.Information);
+
+            ArtifactHttpClientFactory factory = new ArtifactHttpClientFactory(
+            connection.Credentials,
```
Comment (alex-peck) [inline src/Microsoft.VisualStudio.Services.Agent/JobServer.cs:154]: nit: tab these in :)

File: src/Microsoft.VisualStudio.Services.Agent/JobServerQueue.cs [hunk 7]
```diff
@@ -602,10 +609,34 @@ private async Task UploadFile(UploadFileInfo file)
                     // Create the log
                     var taskLog = await _jobServer.CreateLogAsync(_scopeIdentifier, _hubName, _planId, new TaskLog(String.Format(@"logs\{0:D}", file.TimelineRecordId)), default(CancellationToken));
 
-                    // Upload the contents
+
```
Comment (alex-peck) [inline src/Microsoft.VisualStudio.Services.Agent/JobServerQueue.cs:612]: nit: double newline

**Review by alex-peck — APPROVED** [review]
LGTM

**Pull request merged**